metrics/gauge.go

```diff
@@ -2,97 +2,69 @@ package metrics
 
 import "sync/atomic"
 
-// GaugeSnapshot contains a readonly int64.
-type GaugeSnapshot interface {
-	Value() int64
-}
+// GaugeSnapshot is a read-only copy of a Gauge.
+type GaugeSnapshot int64
 
-// Gauge holds an int64 value that can be set arbitrarily.
-type Gauge interface {
-	Snapshot() GaugeSnapshot
-	Update(int64)
-	UpdateIfGt(int64)
-	Dec(int64)
-	Inc(int64)
-}
+// Value returns the value at the time the snapshot was taken.
+func (g GaugeSnapshot) Value() int64 { return int64(g) }
 
 // GetOrRegisterGauge returns an existing Gauge or constructs and registers a
-// new StandardGauge.
-func GetOrRegisterGauge(name string, r Registry) Gauge {
-	if nil == r {
+// new Gauge.
+func GetOrRegisterGauge(name string, r Registry) *Gauge {
+	if r == nil {
 		r = DefaultRegistry
 	}
-	return r.GetOrRegister(name, NewGauge).(Gauge)
+	return r.GetOrRegister(name, NewGauge).(*Gauge)
 }
 
-// NewGauge constructs a new StandardGauge.
-func NewGauge() Gauge {
-	if !Enabled {
-		return NilGauge{}
-	}
-	return &StandardGauge{}
+// NewGauge constructs a new Gauge.
+func NewGauge() *Gauge {
+	return &Gauge{}
 }
 
-// NewRegisteredGauge constructs and registers a new StandardGauge.
-func NewRegisteredGauge(name string, r Registry) Gauge {
+// NewRegisteredGauge constructs and registers a new Gauge.
+func NewRegisteredGauge(name string, r Registry) *Gauge {
 	c := NewGauge()
-	if nil == r {
+	if r == nil {
 		r = DefaultRegistry
 	}
 	r.Register(name, c)
 	return c
 }
 
-// gaugeSnapshot is a read-only copy of another Gauge.
-type gaugeSnapshot int64
-
-// Value returns the value at the time the snapshot was taken.
-func (g gaugeSnapshot) Value() int64 { return int64(g) }
-
-// NilGauge is a no-op Gauge.
-type NilGauge struct{}
-
-func (NilGauge) Snapshot() GaugeSnapshot { return (*emptySnapshot)(nil) }
-func (NilGauge) Update(v int64)          {}
-func (NilGauge) UpdateIfGt(v int64)      {}
-func (NilGauge) Dec(i int64)             {}
-func (NilGauge) Inc(i int64)             {}
-
-// StandardGauge is the standard implementation of a Gauge and uses the
-// sync/atomic package to manage a single int64 value.
-type StandardGauge struct {
-	value atomic.Int64
-}
+// Gauge holds an int64 value that can be set arbitrarily.
+type Gauge atomic.Int64
```

Comment:
I don't like that this was changed to be just an `int64`. It doesn't improve the implementation code (have to convert it at every use), and the internals are fully exposed.

It's OK to define a struct with only one field.

Reply:
It was a bit inconsistent they way it is _right now_:
```golang
// StandardCounter is the standard implementation of a Counter and uses the
// sync/atomic package to manage a single int64 value.
type StandardCounter atomic.Int64
```
vs
```golang
// StandardGauge is the standard implementation of a Gauge and uses the
// sync/atomic package to manage a single int64 value.
type StandardGauge struct {
	value atomic.Int64
}
```
So I decided to just make them identical